himbaechel/uarch/gatemate/pack.cc

```diff
@@ -0,0 +1,2138 @@
+        NetInfo *clk_net = cell->getPort(id_CLK);
+        if (clk_net) {
+            if (clk_net->name == ctx->id("$PACKER_GND")) {
+                cell->disconnectPort(id_CLK);
+            } else if (clk_net->name == ctx->id("$PACKER_VCC")) {
+                cell->disconnectPort(id_CLK);
+            } else {
+                if (!global_signals.count(clk_net)) {
+                    cell->movePortTo(id_CLK, target, id_OUT4);
+                    target->params[id_SEL_OUT_CLOCK] = Property(Property::State::S1);
+                    return true;
+                } else {
+                    int index = global_signals[clk_net];
+                    cell->movePortTo(id_CLK, target, ctx->idf("CLOCK%d", index + 1));
+                    target->params[id_OUT_CLOCK] = Property(index, 2);
+                }
+            }
+        }
+        return false;
+    };
+    auto set_in_clk = [&](CellInfo *cell, CellInfo *target) {
+        NetInfo *clk_net = cell->getPort(id_CLK);
+        if (clk_net) {
+            if (clk_net->name == ctx->id("$PACKER_GND")) {
+                cell->disconnectPort(id_CLK);
+            } else if (clk_net->name == ctx->id("$PACKER_VCC")) {
+                cell->disconnectPort(id_CLK);
+            } else {
+                if (!global_signals.count(clk_net)) {
+                    cell->movePortTo(id_CLK, target, id_OUT4);
+                    target->params[id_SEL_IN_CLOCK] = Property(Property::State::S1);
+                } else {
+                    int index = global_signals[clk_net];
+                    cell->movePortTo(id_CLK, target, ctx->idf("CLOCK%d", index + 1));
+                    target->params[id_IN_CLOCK] = Property(index, 2);
+                }
+            }
+        }
+    };
+
+    auto merge_ibf = [&](NetInfo *di_net, CellInfo &ci, bool use_custom_clock) -> bool {
+        CellInfo *dff = (*di_net->users.begin()).cell;
+        if (is_gpio_valid_dff(dff)) {
+            if (!global_signals.count(ci.getPort(id_CLK)) && use_custom_clock) {
+                log_warning("Found DFF %s cell, but not enough CLK signals.\n", dff->name.c_str(ctx));
+                return false;
+            }
+            // We configure both GPIO IN and let router decide
+            ci.params[id_IN1_FF] = Property(Property::State::S1);
+            ci.params[id_IN2_FF] = Property(Property::State::S1);
+            packed_cells.emplace(dff->name);
+            ci.disconnectPort(id_Y);
+            dff->movePortTo(id_Q, &ci, id_DI);
+            set_in_clk(dff, &ci);
+            bool invert = int_or_default(dff->params, id_CLK_INV, 0) == 1;
+            if (invert) {
+                ci.params[id_INV_IN1_CLOCK] = Property(Property::State::S1);
+                ci.params[id_INV_IN2_CLOCK] = Property(Property::State::S1);
+            }
+            return true;
+        } else {
+            log_warning("Found DFF %s cell, but it is not valid for merge.\n", dff->name.c_str(ctx));
```

Comment:
this warning could be  bit clearer - like naming the IO too

Reply:
Updated message